---
title: "Enzyme Commission Number Prediction using Descriptors"
subtitle: "Biological Data Science Curse - CompBio Coimbra"
authors: 'V.Haberl F.Moroff'
bibliography: 'EC.bib'
nocite: '@*'
output:
  ioslides_presentation:
    widescreen: yes
    smaller: no
    logo: www/FCTUC_V_FundoClaro.png
    css: www/io.css
# output: pdf_document
date: "2022-10-03"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,eval = FALSE)
#library(bio3d)
library(plotly)
library(ggplot2)
library(dplyr)
AAsamples<- readRDS("data/TRAINING_EC_SEQ.rds")
AAC_df<-readRDS("data/TRAINING_EC.rds")
AA_DESC_APAAC<- readRDS("data/TRAINING_AA_DESC_APAAC.rds")
AA_DESC_df<-readRDS("data/TRAINING_descriptors.rds")
```

## Goal 🥅 {.build}

Train Machine Learning algorithms on different discrete Protein Descriptors to predict _Enzyme Commission Numbers_ (EC -Number).

 - Sequence     > Structure             > Function
 - Sequence     > (AA_Descriptors + ML) > Function
 
Compare the Accuracy of combinations between different ML approaches and different Descriptors.


## Enzyme Commission Numbers{.smaller}


| Class                | Reaction catalyzed                                                                                                        | Typical reaction                                | Enzyme example                          |
|:-------------|:-------------------------------|:-------------|:-------------|
| EC 1 Oxidoreductases | Oxidation/reduction reactions; transfer of H and O atoms or electrons from one substance to another                       | AH + B → A + BH (reduced) A + O → AO (oxidized) | Dehydrogenase, oxidase                  |
| EC 2 Transferases    | Transfer of a functional groupf from one substance to another. The group may be methyl-, acyl-, amino- or phosphate group | AB + C → A + BC                                 | Transaminase, kinase                    |
| EC 3 Hydrolases      | Formation of two products from a substrate by hydrolysis                                                                  | AB + H2O → AOH + BH                             | Lipase, amylase, peptidase, phosphatase |
| EC 4 Lyases          | Non-hydrolytic addition or removal of groups from substrates. C-C, C-N, C-O or C-S bonds may be cleaved                   | RCOCOOH → RCOH + CO2 or [X-A+B-Y] → [A=B + X-Y] | Decarboxylase                           |
| EC 5 Isomerases      | Intramolecule rearrangement, i.e. isomerization changes within a single molecule                                          | ABC → BCA                                       | Isomerase, mutase                       |
| EC 6 Ligases         | Join together two molecules by synthesis of new C-O, C-S, C-N or C-C bonds with simultaneous breakdown of ATP             | X + Y + ATP → XY + ADP + Pi                     | Synthetase                              |
| EC 7 Translocases    | Catalyse the movement of ions or molecules across membranes or their separation within membranes                          |                                                 | Transporter                             |
(from <https://www.abbexa.com/enzyme-commission-number> *altered*)

## Steps - Training 🪜

```{r,eval=TRUE,echo=FALSE,out.width="100%",fig.align='left'}
DiagrammeR::grViz("digraph G {
rankdir=LR;
splines=false;
	fontname='Helvetica,Arial,sans-serif'
	node [fontname='Helvetica,Arial,sans-serif']
	edge [fontname='Helvetica,Arial,sans-serif']

	subgraph cluster_0 {
	  style=filled;
		color=lightblue;
		node [style=filled,shape= cylinder];
		a0 [label= 'Random AAs from UniProt\n N 2500/EC number' ]
		subgraph cluster_01{
		  label= 'Amino Acid Descriptors'
		  style=''
		  node [shape=invhouse]
			x0 [label= 'AAC: AA-composition\n(protR)', lhead=cluster_01, ltail=cluster_1]
			x1 [label = 'Simple AA Stats\n(seqinR)']
			x2 [label= 'AAC+\nSimple AA Stats']
			x3 [label ='APAAC\n(protR)']
		  {rank=same; x0 x1 x2 x3}
		  color=blue
		}

		label = 'Training Dataset';
	}

	
		subgraph cluster_2 {
		node [style=filled, shape =box];
		m0 [label='Linear Discriminant Analysis\n[LDA]']
    m1 [label='Random forest\n[RF]']
    m2 [label= 'K-nearest neighbors\n[k-NN]'] 
    m3 [label= 'Support vector machine\n[SVM]']
		label = 'Machine Learning Models';
		color=blue
		}
	
		subgraph cluster_4 {
	  style=filled;
		color=lightyellow;
		node [style=filled,shape= cylinder];
		b0[label=' ~4500 Enzymes from\nHomo Sapiens proteom']
		label = 'Evaluation Dataset';
		subgraph cluster_04{
		  label= 'Amino Acid Descriptors'
		  style=''
		  node [shape=invhouse]
			z0 [label= 'AAC: AA-composition\n(protR)', lhead=cluster_01, ltail=cluster_1]
			z1 [label = 'Simple AA Stats\n(seqinR)']
			z2 [label= 'AAC+\nSimple AA Stats']
			z3 [label ='APAAC\n(protR)']
		  {rank=same; x0 x1 x2 x3}
		  color=blue
      }
	}

a0 ->{x0 x1 x2 x3} [lhead='cluster_0', ltail='cluster_01'];
{x0 x1 x2 x3}->{m0 m1 m2 m3}->{z0 z1 z2 z3}
{z0 z1 z2 z3}->b0[dir=back]

}")

```



# Creation of the training dataset 💾

## Clone accession numbers from github{.smaller}
```{bash, eval= FALSE}
git clone --depth 1 --filter=blob:none --sparse https://github.com/cansyl/ECPred
cd ECPred
git sparse-checkout set ECPred\ Datasets
cd ./ECPred\ Datasets
rm -r -f !\(EC_Main*\)  # only keep main EC number classification
```

```{bash,eval=TRUE}
tree ./ECPred/ECPred\ Datasets
```


## Sample Function{.smaller}

```{r,file="utils/getAAsample.R",echo=TRUE}

```

## Error Catching 🛑 {.smaller}

```{r,file="utils/getUniProt_custom.R",echo=TRUE}

```

## Downloading the data

```{r,echo=TRUE}
path_train ="ECPred/ECPred Datasets/EC_MainClass_PositiveTrain/"
filenames_train<-paste0(path_train,list.files(path_train))
## Get Aminoacid sequence list sample
  ### <b>
AAsamples<-sapply(filenames_train, getAAsample,n=SAMPLESIZE)
  ### </b>
```

## The Raw Trainingdata

```{r, echo=TRUE,eval=TRUE}
str(AAsamples,list.len=4,strict.width='cut')
```


## Creating Aminoacid Features

advantage of the discrete form: easy to be treated in statistical prediction, 
disadvantage it is hard to directly incorporate the sequence-order information 


 1. Amino acid composition (*AAC*)
 $$ P = [f_1  f_2... f_{20}]{^T}$$
 $$ f_u (u = 1,2,…,20)$$ 
 2. "AAStats:" pysico-chemical classes (Size, polarity, charge), isoelectric point, mol. weight.
 3. combined  AAC + AAstat Descriptors
 4. Amphiphilic Pseudo Amino Acid Composition (*APse-AAC*)
 $$ P = [p_1,  p_2,...,p_{20},p_{20+1}...p_{20+\lambda}]{^T} (\lambda< L)$$


##  AAC & APse-AAC with _protr_ {.smaller}


```{r,file="utils/createAACmatrix.R",echo=TRUE}

```


```{r,file="utils/createAAPAAC.R",echo=TRUE}

```



--------------

```{r,echo=TRUE,eval=FALSE}
## create aminoacid content matrices
mats <- sapply(AAsamples, createAACmatrix,simplify = FALSE)
## combine matrices to dataframe
AAC_df <- purrr::map_df(mats, ~as.data.frame(.x), .id="EC")
```

```{r, echo=TRUE,eval=TRUE}
str(AAC_df,list.len=20,strict.width='cut')
```


------------

```{r,echo=TRUE,eval=FALSE}
## create APAAC
APAAC<-sapply(AAsamples, createDescriptors_APAAC,simplify = FALSE)
AA_DESC_APAAC <- purrr::map_df(APAAC, ~as.data.frame(.x), .id="EC")

```

```{r, echo=TRUE,eval=TRUE}
str(AA_DESC_APAAC,list.len=20,strict.width='cut')
```

## SeqinR AminoAcid Stats{.smaller}

```{r,file="utils/createDescriptors.R",echo=TRUE}

```

-------------

```{r,echo=TRUE,eval=FALSE}
## create basic Descriptors (seqinr):
AA_DESC<-sapply(AAsamples,createDescriptors,simplify = FALSE)
AA_DESC_df <- purrr::map_df(AA_DESC, ~as.data.frame(.x), .id="EC")
```


```{r, echo=TRUE,eval=TRUE}
str(AA_DESC_df,list.len=20,strict.width='cut')
```



## Evaluation data

```{r,echo=TRUE,eval=FALSE}
EC_codes<-c(
  "Oxidoreductases"=1,"Transferases"=2, "Hydrolases"=3,
  "Lyases"=4, "Isomerases"=5,  "Ligases"=6
  )

Human_prots<-empty.dump()  ## Empty list
for( i in EC_codes){
  url <- paste0("https://rest.uniprot.org/uniprotkb/stream?format=fasta&query=%28%28taxonomy_id%3A9606%29%20AND%20%28ec%3A",
                i,"%29%29%20AND%20%28reviewed%3Atrue%29")
  Human_prots[[names(EC_codes)[i]]]<-  protr::readFASTA(url,seqonly = TRUE) %>% 
    lapply(.,protr::removeGaps) %>%  # remove gaps in seq 
    subset(lapply(.,protr::protcheck)==TRUE)
}
saveRDS(Human_prots,"data/HSapiens_proteom.rds")

```

## Create Descriptors for Human Protein Data 

Same procedure as for training data
```{r echo=TRUE,eval=FALSE}
Human_prots<-readRDS("data/HSapiens_proteom.rds")
## create aminoacid content matrices For Human Proteom
Human_AAC     <- sapply(Human_prots, createAACmatrix,simplify = FALSE)
Human_AAC_df  <- purrr::map_df(Human_AAC, ~as.data.frame(.x), .id="EC")
saveRDS(Human_AAC_df,"data/HSapiens_proteom_AAC_df.rds")

## create basic Descriptors (seqinr):
Human_AA_DESC <- sapply(Human_prots,createDescriptors,simplify = FALSE)
Human_AA_DESC_df <- purrr::map_df(Human_AA_DESC, ~as.data.frame(.x), .id="EC")
saveRDS(Human_AA_DESC_df,"data/HSapiens_proteom_AA_DESC_df.rds")

## Create APAAC Descriptors for Human Proteom:
Human_APAAC   <- sapply(Human_prots, createDescriptors_APAAC,simplify = FALSE)
Human_APAAC_df <- purrr::map_df(Human_APAAC, ~as.data.frame(.x), .id="EC")
saveRDS(Human_APAAC_df,"data/HSapiens_proteom_APAAC_df.rds")

```


## Modeling Methods

We trained and compared the performance of four different ML approaches:

  - Linear Discriminant Analysis [LDA]
  - Random forest [RF]
  - K-nearest neighbors [k-NN] 
  - Support vector machine [SVM]


## lda (Linear discriminant analysis) 📏

- linear model for classification and dimensionality reduction
- initially for two classes, can be generalized for multiple classes
- projects data from a D dimensional feature space down to a lower dimensional space in a way to maximize the variability between the classes and reducing the variability within the classes
- supervised learning algorithm

## lda (Linear discriminant analysis) 📏

![](img/lda_1.png){width=600x}

## lda (Linear discriminant analysis) 📏
 
![](img/lda_2.png){width=750x}


## knn (k-nearest neighbors algorithm) 🐒

- classification and regression tasks
- estimate the conditional distribution of Y given X
- classify a given observation to the class with highest estimated probability

## knn (k-nearest neighbors algorithm) 🐒

- input consists of the k closest training examples in the feature space
- classified by a majority vote of its neighbors
- Given a positive integer K and a test observation x_0, the KNN classifier first identifies the K points in the training data that are closest to x_0
- classifies the test observation x_0 to the class with the largest probability

## knn (k-nearest neighbors algorithm) 🐒

![](img/knn_1.png){width=650x}

## knn (k-nearest neighbors algorithm) 🐒

![](img/knn_2.png){width=650x}

## svm (Support vector machine)🚀

- classification and regression tasks
- kernels to transform the input data into a higher-dimensional space
- find the best possible linear separation between the data points
- create a model that can accurately predict the outcome of new data

## Hyperplane

two dimensional: 
$$\beta_0 + \beta_1X_1 + \beta_2X_2 = 0$$

p-dimensional: 

$$\beta_0 + \beta_1X_1 + \beta_2X_2 +···+ \beta_pX_p = 0$$

## Hyperplane


![](img/svm_1.png){width=600x}

## Hyperplane

![](img/svm_2.png){width=800x}

## Hyperplane

![](img/svm_3.png){width=550x}


## potential problems ⚠️

![](img/svm_4.png){width=800x}

## potential problems ⚠️

![](img/svm_5.png){width=800x}

## rf (random forest) 🏕️

- classification and regression tasks
- meta-algorithm that combines the predictions of several individual learning algorithms
- many individual decision trees trained on a different subset of the data, using a random subset of the features
- combined to make a single prediction

## rf (random forest) 🏕️

![](img/rf_1.png){width=850x} 


## Model training 🏋 {.smaller} 

```{r,echo=TRUE}
# set EC as factor
data_APAAC$EC <- as.factor(data_APAAC$EC)
## Preprocessing
processing <- preProcess(data_APAAC, method = c("center","scale")) 
data_APAAC <- predict(processing, data_APAAC) 

control <- trainControl(method="cv", number=5)
metric <- "Accuracy"

# 4 dataframes with APAAC descriptors as model features
fit.lda_APAAC <- train(EC~., data=data_APAAC, method="lda", metric=metric, trControl=control)
fit.rf_APAAC <- train(EC~., data=data_APAAC, method="rf", metric=metric, trControl=control)
fit.knn_APAAC <- train(EC~., data=data_APAAC, method="knn", metric=metric, trControl=control)
fit.svm_APAAC <- train(EC~., data=data_APAAC, method="svmRadial", metric=metric, trControl=control)

models_APAAC<-list(lda=fit.lda_APAAC, rf=fit.rf_APAAC, knn=fit.knn_APAAC,svm=fit.svm_APAAC)
saveRDS(models_APAAC,"results/models_APAAC.rds")
```


# Model Evaluation 📈 

## Model Evaluation{.smaller}

```{r}

createConfusionMatrices<-function(data,models){
  Matrices<-empty.dump()
  
  for (d in seq_along(data)) {
    message(paste("Data" ,names(data)[[d]],":"))
    
    cmlist<-empty.dump()
    for (mod in names(models[[d]])) {
      message(mod)
      p     <- predict(models[[d]][[mod]], newdata=data[[d]][-1])
      cmlist[[mod]]<- caret::confusionMatrix(data = p, reference=data[[d]]$EC)
    }
   
    Matrices[[d]] <- cmlist
  }
  names(Matrices)<-names(data)
  return(Matrices)
    
}

```

## Model Evaluation{.smaller}

```{r}
All_data<-list(
  "AAC"      = HS_prot_AAC,
  "DESC"     = HS_prot_DESC,
  "AAC_DESC" = HS_prot_merged,
  "APAAC"    = HS_prot_APAAC
  )

All_data<-lapply(All_data,\(x) predict(preProcess(x, method = c("center","scale")),x))
All_models<-list(models_AAC,models_DESC,models_merged,models_APAAC)

conf_Matrices <- createConfusionMatrices(All_data,All_models)

```

-----------
```{r,echo=FALSE,eval=TRUE}
conf_Matrices<-readRDS("results/conf_matrices.rds")
cbind(conf_Matrices$APAAC$knn$overall)|> t()|>kableExtra::kable(caption = "Overall Statistics & Confusion Matrix for APse-AAC Descriptor k-NN")%>%  kableExtra::kable_paper("hover", full_width = F) 
conf_Matrices$APAAC$knn$table|> kableExtra::kable() %>% kableExtra::kable_paper() %>% kableExtra::add_header_above(c(" "=1, "Reference" = 6 ))

# http://www.csbio.sjtu.edu.cn/bioinf/PseAA/type2.htm
  #corrplot::corrplot(type="lower",diag = FALSE,is.corr = FALSE)
```

# Results

## Results

```{r,file="scripts/createBoxplots.R",echo=FALSE,eval=TRUE, out.width="90%"}

```

# **Take Away** 🏄

## Software

- protr_1.6-2
- seqinr_4.2-16
- caret_6.0-93
- R version 4.1.2

## References{.smaller}

<div id="refs"></div>

